volumegroup/volumegroup.proto

```diff
@@ -0,0 +1,181 @@
+}
+// CreateVolumeGroupRequest holds the required information to
+// create a volume group
+message CreateVolumeGroupRequest {
+  // suggested name for volume group (required for idempotency)
+  // This field is REQUIRED.
+  string name = 1;
+
+  // params passed from VolumeGroupClass
+  // This field is OPTIONAL.
+  map<string, string> parameters = 2;
+
+  // Secrets required by plugin to complete volume group creation
+  // request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 3 [(csi.v1.csi_secret) = true];
+}
+// CreateVolumeGroupResponse holds the information to send when
+// volumeGroup is successfully created.
+message CreateVolumeGroupResponse {
+  // Contains all attributes of the newly created volume group.
+  // This field is REQUIRED.
+  VolumeGroup volume_group = 1;
+}
+// Information about a specific volumeGroup.
+message VolumeGroup {
+  // The identifier for this volume group, generated by the plugin.
+  // This field is REQUIRED.
+  string volume_group_id = 1;
+
+  // Opaque static properties of the volume group.
+  // This field is OPTIONAL.
+  map<string, string> volume_group_context = 2;
+
+  // Underlying volumes in this group. The same definition in CSI
+  // Volume.
+  // This field is OPTIONAL.
+  // To support the creation of an empty group, this list can be empty.
+  // However, this field is not empty in the following cases:
+  // - Response from ListVolumeGroups or ControllerGetVolumeGroup if the
+  //   VolumeGroup is not empty.
+  // - Response from ModifyVolumeGroupMembership if the
+  //   VolumeGroup is not empty after modification.
+  repeated csi.v1.Volume volumes = 3;
+}
+// DeleteVolumeGroupRequest holds the required information to
+// delete a volume group
+message DeleteVolumeGroupRequest {
+  // The ID of the volume group to be deleted.
+  // This field is REQUIRED.
+  string volume_group_id = 1;
+
+  // Secrets required by plugin to complete volume group
+  // deletion request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 2 [(csi.v1.csi_secret) = true];
+}
+// DeleteVolumeGroupResponse holds the information to send when
+// volumeGroup is successfully deleted.
+message DeleteVolumeGroupResponse {
+}
+// ModifyVolumeGroupMembershipRequest holds the required
+// information to modify a volume group
+message ModifyVolumeGroupMembershipRequest {
+  // The ID of the volume group to be modified.
+  // This field is REQUIRED.
+  string volume_group_id = 1;
+
+  // Specify volume_ids that will be in the modified volume group.
+  // This list will be compared with the volume_ids in the existing
+  // group.
+  // New ones will be added and missing ones will be removed.
+  // If no volume_ids are provided, all existing volumes will
+  // be removed from the group.
+  // This field is OPTIONAL.
+  repeated string volume_ids = 2;
```

Comment:
Some storage systems can have empty VolumeGroups, or the VolumeGroup needs to be created before volumes can be added to it. Possibly VolumeGroups can not be created dynamically, only by storage admins (with super powers :superhero:). For these cases, the GetVolumeGroup RPC can be used, after which volumes can be added.

Reply:
agree with @nixpanic, and we need the volume_ids in the modify call, because the driver needs to know which volumes to add and which to remove. In delete call, we just deleting the VG we don't care about the volumes in it

Reply:
sounds good 👍🏻